<mat-card class="analysis-card">
  <mat-card-header>
    <div mat-card-avatar class="analysis-card__img"></div>
    <mat-card-title>topGO: {{ outputs?.job_name }}</mat-card-title>
  </mat-card-header>
  <mat-card-content class="analysis-card__main">
    <p class="analysis-card__instruction">
      Use the table to navigate
      the topGO results. <br>
    </p>
    <p>
      In the table, you will see two p-values: "elim" and "classic". The so-called "elim" method is an adjustment
      that attempts to account for redundancy due to the hierarchical structure
      of GO terms. Consider GO:0019886 (<em>antigen processing and presentation of exogenous peptide 
      antigen via MHC class II</em>)
      and GO:0019882 (<em>antigen processing and presentation</em>); the first is clearly a more specific version
      of the latter and the GO hierarchy reflects this. In a given experiment, if many of the same 
      differentially expressed genes
      are shared by both of these terms, one would generally be more interested in the more specific 
      one. The "elim" method thus attempts to capture this notion. For more details, please see
      <a href="https://academic.oup.com/bioinformatics/article/22/13/1600/193669" target="_blank">
        https://academic.oup.com/bioinformatics/article/22/13/1600/193669
      </a> 
    </p>
    <p class="analysis-card__instruction">
      You can create custom gene sets based on the annotated genes of each GO term.
      Note that due to differences in gene symbols and GO databases, there are often
      discrepancies between the "annotated" genes for a particular GO term and those
      that will be exported as part of any feature sets. You may see gene sets that 
      report zero genes.
    </p>

    <p>Accordingly, we also provide an external query to the 
      <a href="http://amigo.geneontology.org/" target="_blank">AmiGO</a> database which will retrieve gene symbols
        corresponding to a particular GO term. Note that the gene symbols which are returned may or 
        may not properly align with the gene identifiers in your file(s).
        For example, if you are using Ensembl ENSG IDs, you will need to map the symbols to the ENSG IDs
        if you wish to do any further analysis (e.g. plot an expression heatmap for the genes associated
        with a particular GO term).
    </p>
    <mat-divider [inset]="true"></mat-divider>
    <div class="analysis-card__content">
      <section class="filter-section">
        <form [formGroup]="filterForm" (ngSubmit)="onSubmit()">
          <div class="filter">
            <span class="label">
              P-value (classic method):
            </span>
            <mat-form-field class="form-control form-control__small" color="accent">
              <mat-select formControlName="classic_pval_operator">
                <mat-option *ngFor="let operator of operators" [value]="operator.id">
                  {{ operator.name }}
                </mat-option>
              </mat-select>
            </mat-form-field>
            <mat-form-field class="form-control">
              <input matInput formControlName="classic_pval" type=number step=any min=0>
            </mat-form-field>
          </div>
          <div class="filter">
            <span class="label">
              P-value (Elim method)
            </span>
            <mat-form-field class="form-control form-control__small" color="accent">
              <mat-select formControlName="elim_pval_operator">
                <mat-option *ngFor="let operator of operators" [value]="operator.id">
                  {{ operator.name }}
                </mat-option>
              </mat-select>
            </mat-form-field>
            <mat-form-field class="form-control">
              <input matInput formControlName="elim_pval" type=number step=any min=0>
            </mat-form-field>
          </div>
          <button type="submit" [disabled]="!filterForm.valid" mat-raised-button color="accent">Apply filters</button>
        </form>

      </section>

      <div class="mat-elevation-z8">
        <table mat-table class="result-table" [dataSource]="dataSource" matSort [matSortActive]="defaultSorting.field"
          [matSortDirection]="defaultSorting.direction" matSortDisableClear>

          <ng-container matColumnDef="go_id">
            <th mat-header-cell *matHeaderCellDef>GO ID</th>
            <td mat-cell *matCellDef="let row">
              <a href="http://amigo.geneontology.org/amigo/term/{{row.go_id}}" target="_blank">
                {{row.go_id}}
              </a>
            </td>
          </ng-container>

          <ng-container matColumnDef="term">
            <th mat-header-cell *matHeaderCellDef mat-sort-header>Term</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.term}}</td>
          </ng-container>

          <ng-container matColumnDef="annotated">
            <th mat-header-cell *matHeaderCellDef mat-sort-header>Annotated genes</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.annotated}}</td>
          </ng-container>

          <ng-container matColumnDef="significant">
            <th mat-header-cell *matHeaderCellDef mat-sort-header># Significant</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.significant}}</td>
          </ng-container>

          <ng-container matColumnDef="expected">
            <th mat-header-cell *matHeaderCellDef># Expected</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.expected}}</td>
          </ng-container>
          
          <ng-container matColumnDef="elim_pval">
            <th mat-header-cell *matHeaderCellDef mat-sort-header>P-val (Elim)</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.elim_pval}}</td>
          </ng-container>

          <ng-container matColumnDef="classic_pval">
            <th mat-header-cell *matHeaderCellDef mat-sort-header>P-val (Classic)</th>
            <td mat-cell class="description-cell" *matCellDef="let row">{{row.classic_pval}}</td>
          </ng-container>


          <ng-container matColumnDef="actions">
            <th mat-header-cell *matHeaderCellDef></th>

            <td class="action-cell" mat-cell *matCellDef="let row; let i=index;">
              <button class="action-btn" mat-raised-button color="accent" title="Create feature set of {{row.genelist.length}}" 
                (click)="onCreateFeatureSet(row)"
                [disabled]="row.genelist.length===0">
                Create feature set ({{row.genelist.length}} genes)
              </button>
              <button class="action-btn" mat-raised-button color="accent" title="Query AmiGO for genes" 
                (click)="createAmigoFeatureSet(row)">
                Query AmiGO
              </button>
            </td>
          </ng-container>

          <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
          <tr mat-row *matRowDef="let row; columns: displayedColumns"></tr>
        </table>
        <mat-paginator [length]="dataSource.goTermsCount" [pageSize]="10" [pageSizeOptions]="[10, 25, 50]">
        </mat-paginator>
      </div>
    </div>

  </mat-card-content>

</mat-card>